src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -382,22 +389,27 @@ public static class Deleter extends ItemListener {
 
         private static /* almost final */ int CLEANUP_THREAD_LIMIT = SystemProperties.getInteger(Deleter.class.getName() + ".CLEANUP_THREAD_LIMIT", Integer.valueOf(0)).intValue();
 
+        private final transient ExecutorService executorService = Executors.newSingleThreadExecutor(threadFactory());
+
         /** Semaphore for limiting number of scheduled {@link CleanupTask} */
-        private static Semaphore cleanupPool = new Semaphore(CLEANUP_THREAD_LIMIT, true);
+        private static final Semaphore cleanupPool = new Semaphore(CLEANUP_THREAD_LIMIT, true);
 
         /** Number of {@link CleanupTask} which have been scheduled but not yet completed. */
         private static int runningTasks;
 
+        private static ThreadFactory threadFactory() {
+            return new NamingThreadFactory(new DaemonThreadFactory(), "DeleterCleanupTask");
+        }
```

Comment:
If you go down the route of creating a new executor service I think you need quite a few more modifiers to avoid various problems: https://github.com/jenkinsci/jenkins/blob/d22cc2fa3c426420514240e351faa56ea12e5446/core/src/main/java/hudson/model/Computer.java#L1348-L1353

With the current approach of a single thread executor to dispatch the tasks, it might be simpler to instead use `new SingleLaneExecutorService(Computer.threadPoolForRemoting)` since we're not using the new thread pool for the actual `CleanupTask`s anyway.

> Or was also wondering if Computer.threadPoolRemoting was required here ? Why not directly use a specific thread pool with unbounded queue (initial thread to 1, max number of threads to the limit set, 60s keep alive time) ?

This idea seems fine too and I think would allow us to get rid of all the semaphores to simplify things, right? The only thing I would consider changing would be to make it so that if `CLEANUP_THREAD_LIMIT` is 0, continue to use `Computer.threadPoolForRemoting` directly to avoid creating an extra unbounded thread pool. You'd no longer be able to edit `CLEANUP_THREAD_LIMIT` without restarting Jenkins, but that seems fine.

Reply:
> This idea seems fine too and I think would allow us to get rid of all the semaphores to simplify things, right? The only thing I would consider changing would be to make it so that if CLEANUP_THREAD_LIMIT is 0, continue to use Computer.threadPoolForRemoting directly to avoid creating an extra unbounded thread pool. You'd no longer be able to edit CLEANUP_THREAD_LIMIT without restarting Jenkins, but that seems fine.

Allright, I'll propose something like that then.

Reply:
Per my understanding of environment where this happen (and branch indexing in general) the expected behavior is to see spikes of CleanupTasks at irregular interval during the day, rather than seeing a constant queuing of them.

So I propose `ThreadPoolExecutor` with a `corePoolSize` at `0` and setting the max per `CLEANUP_THREAD_LIMIT` with `SynchronousQueue` so that the thread pool can react quickly to the load, spawning up to the maximum number of threads and eventually retiring the 10s idle threads until the next batch.